aws/hll/aws-s3-transfer-manager/src/download/discovery.rs

```diff
@@ -0,0 +1,337 @@
+                    *r.start(),
+                    cmp::min(*r.start() + handle.target_part_size - 1, *r.end()),
+                ),
+                None => ByteRange::Inclusive(0, handle.target_part_size - 1),
+            };
+            let r = request
+                .input
+                .clone()
+                .set_part_number(None)
+                .range(header::Range::bytes(byte_range));
+
+            discover_obj_with_get(handle, r, range).await
+        }
+    }
+}
+
+async fn discover_obj_with_head(
+    handle: &DownloadHandle,
+    request: &DownloadRequest,
+    byte_range: Option<ByteRange>,
+) -> Result<ObjectDiscovery, error::TransferError> {
+    let meta: ObjectMetadata = handle
+        .client
+        .head_object()
+        .set_bucket(request.input.get_bucket().clone())
+        .set_key(request.input.get_key().clone())
+        .send()
+        .await
+        .map_err(|e| error::DownloadError::DiscoverFailed(e.into()))?
+        .into();
+
+    let remaining = match byte_range {
+        Some(range) => match range {
+            ByteRange::Inclusive(start, end) => start..=end,
+            ByteRange::AllFrom(start) => start..=meta.total_size(),
+            ByteRange::Last(n) => (meta.total_size() - n + 1)..=meta.total_size(),
+        },
+        None => 0..=meta.total_size(),
+    };
+
+    Ok(ObjectDiscovery {
+        remaining,
+        meta,
+        initial_chunk: None,
+    })
+}
+
+async fn discover_obj_with_get(
+    handle: &DownloadHandle,
+    request: GetObjectInputBuilder,
+    range: Option<RangeInclusive<u64>>,
+) -> Result<ObjectDiscovery, error::TransferError> {
+    let resp = request.send_with(&handle.client).await;
+
+    if resp.is_err() {
+        // TODO - deal with empty file errors, see https://github.com/awslabs/aws-c-s3/blob/v0.5.7/source/s3_auto_ranged_get.c#L147-L153
+    }
+
+    let mut resp = resp.map_err(|e| error::DownloadError::DiscoverFailed(e.into()))?;
+
+    // take the body so we can convert the metadata
+    let empty_stream = ByteStream::new(SdkBody::empty());
+    let body = mem::replace(&mut resp.body, empty_stream);
+
+    let data = body
+        .collect()
+        .await
+        .map_err(|e| error::DownloadError::DiscoverFailed(e.into()))?;
+
+    let meta: ObjectMetadata = resp.into();
+
+    let remaining = match range {
+        Some(range) => (*range.start() + data.remaining() as u64)..=*range.end(),
+        None => (data.remaining() as u64)..=meta.total_size() - 1,
+    };
+
+    Ok(ObjectDiscovery {
+        remaining,
+        meta,
+        initial_chunk: Some(data),
+    })
+}
+
+#[cfg(test)]
+mod tests {
+    use crate::{
+        download::{
+            discovery::{
+                discover_obj, discover_obj_with_get, discover_obj_with_head,
+                ObjectDiscoveryStrategy,
+            },
+            handle::DownloadHandle,
+            header::ByteRange,
+        },
+        MIN_PART_SIZE,
+    };
```

Comment:
I love doing imports like this but it's not in keeping with the style of imports in our other crates. Consider updating this to match our established style.

Reply:
I think this is all auto imported...will have to sort out what IDE setting to change 🤔

Reply:
we don't seem to have a rustfmt configuration defined anywhere, this could be controlled via https://rust-lang.github.io/rustfmt/?version=master&search=import#imports_granularity